File: src/main/java/keiproductfamily/rtmAddons/trainDetector/EntityTrainDetectorAdvance.java
package keiproductfamily.rtmAddons.trainDetector;

import jp.ngt.rtm.RTMItem;
import jp.ngt.rtm.electric.EntityElectricalWiring;
import jp.ngt.rtm.electric.SignalLevel;
import jp.ngt.rtm.entity.train.EntityBogie;
import jp.ngt.rtm.entity.train.EntityTrainBase;
import jp.ngt.rtm.item.ItemInstalledObject.IstlObjType;
import jp.ngt.rtm.item.ItemWire;
import jp.ngt.rtm.rail.TileEntityLargeRailBase;
import keiproductfamily.GuiIDs;
import keiproductfamily.ModKEIProductFamily;
import keiproductfamily.rtmAddons.RTMAChannelData;
import keiproductfamily.rtmAddons.RTMAChannelMaster;
import net.minecraft.entity.Entity;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.world.World;

import java.util.List;
import java.util.Objects;
import java.util.stream.IntStream;

public class EntityTrainDetectorAdvance extends EntityElectricalWiring {
    private boolean findTrain;
    private String _channelName = "";
    private int _channelNumber = 0;

    public EntityTrainDetectorAdvance(World world) {
        super(world);
        this.setSize(1.0F, 0.0625F);
        this.ignoreFrustumCheck = true;
    }

    @Override
    protected void entityInit() {
        super.entityInit();
    }

    @Override
    public void writeEntityToNBT(NBTTagCompound nbt) {
        super.writeEntityToNBT(nbt);
        nbt.setString("channelName", _channelName);
        nbt.setInteger("channelNumber", _channelNumber);
    }

    @Override
    public void readEntityFromNBT(NBTTagCompound nbt) {
        super.readEntityFromNBT(nbt);
        this._channelName = nbt.getString("channelName");
        this._channelNumber = nbt.getInteger("channelNumber");
    }

    public String getChannelName() {
        return _channelName;
    }

    public int getChannelNumber() {
        return _channelNumber;
    }

    public String getChannelKey() {
        return _channelName + _channelNumber;
    }

    public void setChunnelName(String chunnelName, int channelNumber) {
        chunnelName = chunnelName.trim();
        if (!_channelName.equals(chunnelName) || _channelNumber != channelNumber) {
            if (!chunnelName.equals("")) {
                this._channelName = chunnelName;
                this._channelNumber = channelNumber;
            }
        }
    }

    @Override
    public void onUpdate() {
        if (!this.worldObj.isRemote) {
            if (_channelNumber > 0) {
                RTMAChannelData channelData = RTMAChannelMaster.getChannelData(getChannelKey());
                if (channelData != null) {
                    EntityTrainBase train = getCollidedTrain();
                    if (train != null) {
                        byte rollSignID = train.getTrainStateData(8);
                        channelData.setTrainData(getSignalLevel(), rollSignID);
                    } else {
                        channelData.setTrainData(getSignalLevel(), (byte) -1);
                    }
                }
            }
        }

        super.onUpdate();
    }

    public EntityTrainBase getCollidedTrain() {
        List collideEntityList = this.worldObj.getEntitiesWithinAABBExcludingEntity(this, this.getBoundingBox().expand(0, 2, 0));
        if (collideEntityList != null) {
            for (Object entity : collideEntityList) {
                if (entity instanceof EntityTrainBase) {
                    findTrain = true;
                    return ((EntityTrainBase) entity).getFormation().get(0).train;
                }
                if (entity instanceof EntityBogie) {
                    findTrain = true;
                    return ((EntityBogie) entity).getTrain().getFormation().get(0).train;
                }
            }
        }
        findTrain = false;
        return null;
    }

    @Override
    public boolean interactFirst(EntityPlayer player) {
        if (player.getHeldItem() != null || !(player.getHeldItem().getItem() instanceof ItemWire)) {
            if (worldObj.isRemote) {
                player.openGui(ModKEIProductFamily.instance, GuiIDs.GuiID_EntityTrainDetectorSetting, worldObj, this.getEntityId(), 0, 0);
            }
            return true;
        }
        return super.interactFirst(player);
    }

    @Override
    public int getElectricity() {
        return this.findTrain ? SignalLevel.STOP.level : SignalLevel.PROCEED.level;
    }

    private SignalLevel getSignalLevel() {
        return this.findTrain ? SignalLevel.STOP : SignalLevel.PROCEED;
    }

    @Override
    public void setElectricity(int par1) {
    }

    @Override
    protected void dropItems() {
        //Todo
        this.entityDropItem(new ItemStack(ModKEIProductFamily.itemTrainDetectorAdvance), 0.0F);
    }

    @Override
    public String getSubType() {
        return "Antenna_Receive";
    }

    @Override
    protected String getDefaultName() {
        return "TrainDetector_01";
    }

    @Override
    protected ItemStack getItem() {
        return new ItemStack(ModKEIProductFamily.itemTrainDetectorAdvance);
    }
}

File: src/main/java/keiproductfamily/rtmAddons/trainDetector/EntityTrainDetectorAdvance.kt
package keiproductfamily.rtmAddons.trainDetector

import jp.ngt.rtm.electric.EntityElectricalWiring
import jp.ngt.rtm.electric.SignalLevel
import jp.ngt.rtm.entity.train.EntityBogie
import jp.ngt.rtm.entity.train.EntityTrainBase
import jp.ngt.rtm.entity.train.parts.EntityFloor
import jp.ngt.rtm.item.ItemWire
import keiproductfamily.*
import keiproductfamily.PermissionList.IParmission
import keiproductfamily.network.PacketHandler
import keiproductfamily.rtmAddons.ChannelKeyPair
import keiproductfamily.rtmAddons.RequestEntityNBTData
import keiproductfamily.rtmAddons.detectorChannel.EnumDirection
import keiproductfamily.rtmAddons.detectorChannel.RTMDetectorChannelMaster
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.util.AxisAlignedBB
import net.minecraft.world.World
import kotlin.math.abs
import kotlin.math.atan2
import kotlin.math.min

class EntityTrainDetectorAdvance(world: World?) : EntityElectricalWiring(world), IParmission {
    private var findTrain = false
    var channelKeyPair: ChannelKeyPair = ChannelKeyPair("", "")

    override fun entityInit() {
        super.entityInit()
    }

    override fun setEntityId(p_145769_1_: Int) {
        super.setEntityId(p_145769_1_)
        if (worldObj.isRemote) {
            PacketHandler.sendPacketServer(RequestEntityNBTData(this))
        }
    }

    public override fun writeEntityToNBT(nbt: NBTTagCompound) {
        super.writeEntityToNBT(nbt)
        nbt.setChannelKeyPair("channelKeyPair", channelKeyPair)
    }

    public override fun readEntityFromNBT(nbt: NBTTagCompound) {
        channelKeyPair = nbt.getChannelKeyPair("channelKeyPair")
    }

    val channelKey: String
        get() = channelKeyPair.keyString

    fun setChannelName(channelName: String, channelNumber: String) {
        val name = channelName.trim { it <= ' ' }
        val number = channelNumber.trim { it <= ' ' }
        if (channelKeyPair.name != name || channelKeyPair.number != number) {
            if (name != "" && number != "") {
                channelKeyPair = ChannelKeyPair(name, number)
                RTMDetectorChannelMaster.makeChannel(channelKeyPair)
            }
        }
    }

    var lastCollidedTrainFormationID: Long = -1
    var lastCollidedTrainFormationIDKeepCnt: Int = -1
    override fun onUpdate() {
        if (!worldObj.isRemote) {
            if (channelKeyPair.number != "") {
                val channelData = RTMDetectorChannelMaster.getChannelData(channelKey)
                if (channelData != null) {
                    val (train, direction) = getCollidedTrainData()
                    if (train != null) {
                        if (lastCollidedTrainFormationID != train.formation.id) {
                            val rollSignID = train.getTrainStateData(8)
                            lastCollidedTrainFormationID = train.formation.id
                            channelData.setTrainData(
                                ModCommonVar.findTrainLevel,
                                rollSignID,
                                lastCollidedTrainFormationID,
                                direction
                            )
                        }
                        lastCollidedTrainFormationIDKeepCnt = 200
                    } else {
                        channelData.setTrainData(ModCommonVar.notfindTrainLevel, (-1).toByte(), -1L, direction)
                        if (lastCollidedTrainFormationIDKeepCnt > 0) {
                            lastCollidedTrainFormationIDKeepCnt--
                        } else if (lastCollidedTrainFormationIDKeepCnt == 0) {
                            lastCollidedTrainFormationID = -1
                            lastCollidedTrainFormationIDKeepCnt = -1
                        }
                    }
                }
            }
        }
        super.onUpdate()
    }

    override fun getBoundingBox(): AxisAlignedBB {
        return boundingBox
    }

    fun getCollidedTrainData(): Pair<EntityTrainBase?, EnumDirection> {
        if (worldObj == null) {
            return Pair(null, EnumDirection.Null)
        }
        val collideEntityList =
            worldObj.getEntitiesWithinAABBExcludingEntity(this, getBoundingBox().expand(0.0, 2.0, 0.0))
        if (collideEntityList != null) {
            for (entity in collideEntityList) {
                if (entity is EntityTrainBase) {
                    findTrain = true
                    val angYaw = Math.toDegrees(atan2(entity.posX - entity.prevPosX, entity.posZ - entity.prevPosZ))
                    val diff0 = (angYaw - this.rotationYaw) % 360
                    val diff1 = (angYaw - this.rotationYaw + 360) % 360
                    val direction = if (min(abs(diff0), abs(diff1)) < 90) {
                        EnumDirection.Elimination
                    } else {
                        EnumDirection.Access
                    }
                    return Pair(entity.formation[0]?.train, direction)
                }
                if (entity is EntityBogie) {
                    findTrain = true
                    val diff0 = (entity.movingYaw - this.rotationYaw) % 360
                    val diff1 = (entity.movingYaw - this.rotationYaw + 360) % 360
                    val direction = if (min(abs(diff0), abs(diff1)) < 90) {
                        EnumDirection.Elimination
                    } else {
                        EnumDirection.Access
                    }
                    return Pair(entity.train?.formation?.get(0)?.train, direction)
                }
                if (entity is EntityFloor) {
                    val vehicle = entity.vehicle
                    if (vehicle is EntityTrainBase) {
                        findTrain = true
                        val angYaw = Math.toDegrees(atan2(entity.posX - entity.prevPosX, entity.posZ - entity.prevPosZ))
                        val diff0 = (angYaw - this.rotationYaw) % 360
                        val diff1 = (angYaw - this.rotationYaw + 360) % 360
                        val direction = if (min(abs(diff0), abs(diff1)) < 90) {
                            EnumDirection.Elimination
                        } else {
                            EnumDirection.Access
                        }
                        return Pair(vehicle.formation[0]?.train, direction)
                    }
                }
            }
        }
        findTrain = false
        return Pair(null, EnumDirection.Null)
    }

    override fun interactFirst(player: EntityPlayer): Boolean {
        if (player.heldItem == null || player.heldItem.item !is ItemWire) {
            if (worldObj.isRemote) {
                player.openGui(
                    ModKEIProductFamily.instance, GuiIDs.GuiID_EntityTrainDetectorSetting, worldObj,
                    entityId, 0, 0
                )
            }
            return true
        }
        return super.interactFirst(player)
    }

    override fun getElectricity(): Int {
        return if (findTrain) ModCommonVar.findTrainLevel.level else ModCommonVar.notfindTrainLevel.level
    }

    private val signalLevel: SignalLevel
        get() = if (findTrain) ModCommonVar.findTrainLevel else ModCommonVar.notfindTrainLevel

    override fun setElectricity(par1: Int) {}
    override fun dropItems() {
        entityDropItem(ItemStack(ModKEIProductFamily.itemTrainDetectorAdvance), 0.0f)
    }

    override fun getSubType(): String {
        return "Antenna_Receive"
    }

    override fun getDefaultName(): String {
        return "TrainDetector_01"
    }

    override fun getItem(): ItemStack {
        return ItemStack(ModKEIProductFamily.itemTrainDetectorAdvance)
    }

    init {
        setSize(1.0f, 0.0625f)
        ignoreFrustumCheck = true
    }

    override fun getName(): String {
        return "EntityTrainDetectorAdvance"
    }
}